sdk/servicebus/service-bus/test/browserified/batchReceiver.spec.ts

```diff
@@ -0,0 +1,719 @@
+  expectedPeekLength: number
+): Promise<void> {
+  const peekedMsgs = await client.peek(expectedPeekLength + 1);
+  should.equal(
+    peekedMsgs.length,
+    expectedPeekLength,
+    "Unexpected number of msgs found when peeking"
+  );
+}
+
+let ns: ServiceBusClient;
+
+let errorWasThrown: boolean;
+
+let senderClient: QueueClient | TopicClient;
+let receiverClient: QueueClient | SubscriptionClient;
+let deadLetterClient: QueueClient | SubscriptionClient;
+let sender: Sender;
+let receiver: Receiver | SessionReceiver;
+const maxDeliveryCount = 10;
+
+async function beforeEachTest(
+  senderType: TestClientType,
+  receiverType: TestClientType,
+  useSessions?: boolean
+): Promise<void> {
+  // The tests in this file expect the env variables to contain the connection string and
+  // the names of empty queue/topic/subscription that are to be tested
+
+  // @ts-ignore
+  if (!window.__env__["SERVICEBUS_CONNECTION_STRING"]) {
+    throw new Error(
+      "Define SERVICEBUS_CONNECTION_STRING in your environment before running integration tests."
+    );
+  }
+
+  // @ts-ignore
+  ns = ServiceBusClient.createFromConnectionString(window.__env__["SERVICEBUS_CONNECTION_STRING"]);
+
+  const clients = await getSenderReceiverClients(ns, senderType, receiverType);
+  senderClient = clients.senderClient;
+  receiverClient = clients.receiverClient;
+  if (receiverClient instanceof QueueClient) {
+    deadLetterClient = ns.createQueueClient(
+      QueueClient.getDeadLetterQueuePath(receiverClient.entityPath)
+    );
+  }
+
+  if (receiverClient instanceof SubscriptionClient) {
+    deadLetterClient = ns.createSubscriptionClient(
+      TopicClient.getDeadLetterTopicPath(senderClient.entityPath, receiverClient.subscriptionName),
+      receiverClient.subscriptionName
+    );
+  }
+
+  await purge(receiverClient, useSessions ? TestMessage.sessionId : undefined);
+  await purge(deadLetterClient);
+  const peekedMsgs = await receiverClient.peek();
+  const receiverEntityType = receiverClient instanceof QueueClient ? "queue" : "topic";
+  if (peekedMsgs.length) {
+    chai.assert.fail(`Please use an empty ${receiverEntityType} for integration testing`);
+  }
+  const peekedDeadMsgs = await deadLetterClient.peek();
+  if (peekedDeadMsgs.length) {
+    chai.assert.fail(
+      `Please use an empty dead letter ${receiverEntityType} for integration testing`
+    );
+  }
+
+  sender = senderClient.createSender();
+  if (useSessions) {
+    receiver = receiverClient.createReceiver(ReceiveMode.peekLock, {
+      sessionId: TestMessage.sessionId
+    });
+  } else {
+    receiver = receiverClient.createReceiver(ReceiveMode.peekLock);
+  }
+}
+
+async function afterEachTest(): Promise<void> {
+  await ns.close();
+}
+describe("Batch Receiver - Settle message", function(): void {
+  afterEach(async () => {
+    await afterEachTest();
+  });
+
+  async function sendReceiveMsg(testMessages: SendableMessageInfo): Promise<ServiceBusMessage> {
```

Comment:
I am assuming most of these test functions (sendReceiveMsg, testComplete etc in this file) are the same as in the correponding test file for node.
If so, then it would be great if we could share the code instead of copying over the code.

Have we tried exporting these functions from say `test/batchRecevier/spec.ts` and importing them here?

Reply:
The tests are not all uniformly structured i.e., nesting of functions etc. seems problematic with some tests.
I would suggest refactoring of tests can be done all at once for better maintainability, after other concerns with current issue is resolved.

Reply:
It gets simpler to track exactly what tests are we running for browser and review the code, if we re-use the code from the node tests.

I agree that there are other concerns in this PR that are being taken care of. In that case, I would suggest to update this PR to have just 1 test (pick any). We use that as a vehicle to resolve all other issues and concerns brought up in this PR regarding running browser tests. 

Then in separate PRs, we can take 1 node test file at a time, refactor as necessary so that the code is re-usable between node and browser.

Reply:
We'll need to refactor the node tests and move out commonly used functions that are repeated across test files, such as peekLength() among others I think.
And we'll also need to address the nesting so that the test specific functions are not private and are marked as exported. ALl of these would be making changes to node tests. So reviewing will be harder as node tests will need to be reviewed, verified as well.

Agree about refactoring and code re-usability. It's going to take much longer to update each test file if the node tests are not already refactored.

Reply:
I would prefer if we reduce the code duplication here like @ramya-rao-a suggested.
If we have to make a fix to a test (or change it because the API has changed) we will have to make the change in multiple test files. Making it such that the code is shared (as much as possible), while doesn't totally remove that risk, it minimizes it.

Let's split this into 2 PRs:
 - One that will bring over the infrastructure for running tests and have 1 test enable and running to prove that everything works.
 - Another that refactors the code to allow for better code sharing and maintenance. Let's file an issue to track this work.

Reply:
Discussed offline about this.
- https://github.com/Azure/azure-sdk-for-js/issues/2859 has been filed to track refactoring of tests.
- As for removing all browser tests and keeping it to just one test, we already have 2 issues surfaced due to running of the tests - https://github.com/Azure/azure-sdk-for-js/issues/2860 and https://github.com/Azure/azure-sdk-for-js/issues/2861. It may be better to still have them all in for easier collaboration, or do we still want to keep to just 1 test?

@AlexGhiondea @ramya-rao-a

Reply:
Let's merge this as-is, address the uncovered issues and then refactor the tests.

Reply:
My understanding was that the set of tests that we added was the subset of the tests we wanted to add (per the 1st checkbox in #2163).

https://github.com/Azure/azure-sdk-for-js/issues/2163#issuecomment-491172859 contains the list of tests that @ramya-rao-a identified as being different between node and browser. That will reduce the duplicated code, makes the code easier to refactor in the next PR as well as will make the tests run a little bit faster (since there are fewer tests that need to run).

Let's trim down the list of tests to the set that will test the different behavior between browser and NodeJS and then let's get this merged in.

Reply:
The current changes in PR add tests based on initial analysis outlined as below from 5/9 which addresses the 1st checkbox in https://github.com/Azure/azure-sdk-for-js/issues/2163.
```
•	No need to have tests for both partitioned and unpartitioned (because our code doesn't differentiate between the 2), just pick one of the two
•	For receiving, we should have tests for both sessions and non-sessions because we have very different code paths for the two.
•	Since Queues and Topic/Subscriptions have same but repeated code, I am ok doing just the Queues. Use Subscriptions for the topic filter related tests
•	I wouldn't keep stress/perf tests for Browser in scope here, especially keeping our June timeline in mind

```

Based on https://github.com/Azure/azure-sdk-for-js/issues/2163#issuecomment-491172859 from 5/10 or 5/11 that came in after these changes were already made, will need to go through each of the test files again. 

Also, it's not clear why we want to exclude tests that test 'service only' from being covered in browser. Shouldn't these be excluded from the SDK itself? 

Also, looks like other refactoring pertaining to errors/error messages are in progress which is requiring manual changes to these tests as well. 
Some of the tests are still relying on error message string comparisons. Those need to be fixed as well as it makes things harder to maintain.

@ramya-rao-a @AlexGhiondea

Reply:
> Also, it's not clear why we want to exclude tests that test 'service only' from being covered in browser. Shouldn't these be excluded from the SDK itself?

Ideally yes. But from what you might have seen during our efforts in testing Service Bus in the past few months, we uncovered quite a few service side errors by having tests that test such service side features. Therefore having such tests part of our test suite were definitely helpful.

When we think of testing the browser version of our sdk, the intention of the test suite is to uncover issues that might occur from the fact that the javascript code is running in the browser and not node. Therefore, it is very essential to cover the business logic we have in our sdk, but not so much the service side logic.

Once the request goes out of our library, the service doesnt care whether the request came from browser or node. Therefore, testing the service side features twice is not needed.

For example:
In browser mode, testing that the message we received can be completed, abandoned, deferred and deadlettered is essential because we have different code paths in our library as well as our dependent libraries (rhea) for each of these features.
But, abandon a message 10 times to test that it lands in the dead letter queue is testing the service side feature. Testing this feature in node is enough to ensure that the service behaves as expected. Testing this feature again in browser doesnt give us any added benefits.

By pruning such cases, we reduce the total number of tests in our test suites that helps in avoiding an unnecessary increase in the test duration. Not only that, it reduces points of failures due to external factors as well. For example, we still have test runs that fail sometimes due to timeout. Today, we go through these test failures to identify if they were caused by code changes or external factors. By adding browser tests for features that are already covered in the node tests, we increase the chances of such things happening resulting in more developer time being spent in chasing errors. 

> Also, looks like other refactoring pertaining to errors/error messages are in progress which is requiring manual changes to these tests as well.

This has been stabilized. The last set of changes went in last week in order to prepare for the GA release.

> Some of the tests are still relying on error message string comparisons. Those need to be fixed as well as it makes things harder to maintain.

Please log an issue enumerating such occurrences with proposed solution for each case.